apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts

```diff
@@ -3,20 +3,104 @@ import { OpenAI } from 'openai';
 import { loadSqlEquiv, queryRows } from '@prairielearn/postgres';
 
 import { QuestionGenerationContextEmbeddingSchema } from '../../lib/db-types.js';
-import { createServerJob } from '../../lib/server-jobs.js';
+import { ServerJob, createServerJob } from '../../lib/server-jobs.js';
 
 import { createEmbedding, openAiUserFromAuthn, vectorToString } from './contextEmbeddings.js';
 
 const sql = loadSqlEquiv(import.meta.url);
 
 /**
- * Generates the HTML and Python code for a new question using an LLM
+ * Generates the common preamble with general PrairieLearn information for the LLM
  *
- * @param client the OpenAI client to use
- * @param courseId the ID of the current course
- * @param authnUserId the authenticated user's ID
- * @param prompt the prompt for how to generate a question
- * @returns a server job ID for the generation task
+ * @param context Relevant example documents, formatted into one string.
+ * @returns A string, the prompt preamble.
+ */
+function promptPreamble(context: string): string {
+  return `# Introduction
+        
+        You are an assistant that helps instructors write questions for PrairieLearn.
+        
+        A question has a \`question.html\` file that can contain standard HTML, CSS, and JavaScript. It also includes PrairieLearn elements like \`<pl-multiple-choice>\` and \`<pl-number-input>\`.
+        
+        A question may also have a \`server.py\` file that can randomly generate unique parameters and answers, and which can also assign grades to student submissions. \`server.py\` may be omitted if it's not necessary.
+        
+        ## Generating random parameters
+        
+        \`server.py\` may define a \`generate\` function. \`generate\` has a single parameter \`data\` which can be modified by reference. It has the following properties:
+        
+        - \`params\`: A dictionary. Random parameters, choices, etc. can be written here for later retrieval.
+        
+        ## Using random parameters
+        
+        Parameters can be read in \`question.html\` with Mustache syntax. For instance, if \`server.py\` contains \`data["params"]["answer"]\`, it can be read with \`{{ params.answer }}\` in \`question.html\`.
+        
+        # Context
+        
+        Here is some context that may help you respond to the user. This context may include example questions, documentation, or other information that may be helpful.
+        
+        ${context}
+        
+        `;
+}
+
+/**
+ * Builds the context string, consisting of relevant documents.
```

Comment:
Add extra blank line before params.

Reply:
Done in 5af63ea5a9d690527f5feb43c89b6d6ea7c7ba30